File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        int ans = -1, n = bloomDay.length;
        if(n >= k*m){
            int low , high, mid;
            boolean target;
            low = min(bloomDay,n);
            high = max(bloomDay,n);
            while(low <= high){
                mid = (low+high)/2;
                target = check(bloomDay,mid,m,k,n);
                if(target){
                    ans = mid;
                    high = mid-1;
                }else{
                    low = mid+1;
                }
            }
        }
        return ans;
    }
    public boolean check(int[] bd,int mid,int m,int k,int n){
        int count=0,c=0;
        for(int i=0;i<n;i++){
            if(bd[i] > mid){
                c = c+(count/k);
                count = 0;
            }else{
                count++;
            }
        }
        c = c+(count/k);
        if(c>=m){
            return true;
        }else{
            return false;
        }
    }
    
    public int max(int[] bD, int n){
        int maxi = Integer.MIN_VALUE;
        for(int i=0;i<n;i++){
            if(maxi < bD[i])
                maxi = bD[i];
        }
        return maxi;
    }
    public int min(int[] bD, int n){
        int mini = Integer.MAX_VALUE;
        for(int i=0;i<n;i++){
            if(mini > bD[i])
                mini = bD[i];
        }
        return mini;
    }
}

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.go
func minDays(bloomDay []int, m int, k int) int {
    ans := -1
    n := len(bloomDay)
    if(n >= k*m){
        low := min(bloomDay,n)
        high := max(bloomDay,n)
        var mid,target int
        for low<=high{
            mid = (low+high)/2
            target = check(bloomDay,mid,k,n)
            if(target >= m){
                ans = mid
                high = mid-1
            }else{
                low = mid+1
            }
        }
    }
    return ans
}

func check(bd []int,mid int,k int,n int) int{
    counter:=0
    count := 0
    for i:=0;i<n;i++{
        if bd[i] > mid{
            count = count + (counter/k)
            counter = 0
        }else{
            counter++
        }
    }
    count = count + (counter/k)
    return count
}

func max(bd []int,n int) int{
    maxi := math.MinInt64
    for i:=0;i<n;i++{
        if(maxi < bd[i]){
            maxi = bd[i]
        }
    }
    return maxi
}

func min(bd []int,n int) int{
    mini := math.MaxInt64
    for i:=0;i<n;i++{
        if(mini > bd[i]){
            mini = bd[i]
        }
    }
    return mini
}